proto/src/serialize/binary/encoder.rs

```diff
@@ -163,59 +159,44 @@ impl<'a> BinEncoder<'a> {
 
     /// Writes a u16 in network byte order to the buffer
     pub fn emit_u16(&mut self, data: u16) -> ProtoResult<()> {
-        self.buffer.reserve(2); // two bytes coming
-
-        let b1: u8 = (data >> 8 & 0xFF) as u8;
-        let b2: u8 = (data & 0xFF) as u8;
-
-        self.emit(b1)?;
-        self.emit(b2)?;
-
+        let mut bytes = [0; 2];
+        {
+            NetworkEndian::write_u16(&mut bytes, data);
+        }
+        self.write_slice(&bytes);
         Ok(())
     }
 
     /// Writes an i32 in network byte order to the buffer
     pub fn emit_i32(&mut self, data: i32) -> ProtoResult<()> {
-        self.buffer.reserve(4); // four bytes coming...
-
-        let b1: u8 = (data >> 24 & 0xFF) as u8;
-        let b2: u8 = (data >> 16 & 0xFF) as u8;
-        let b3: u8 = (data >> 8 & 0xFF) as u8;
-        let b4: u8 = (data & 0xFF) as u8;
-
-        self.emit(b1)?;
-        self.emit(b2)?;
-        self.emit(b3)?;
-        self.emit(b4)?;
-
+        let mut bytes = [0; 4];
+        {
+            NetworkEndian::write_i32(&mut bytes, data);
+        }
+        self.write_slice(&bytes);
         Ok(())
     }
 
     /// Writes an u32 in network byte order to the buffer
     pub fn emit_u32(&mut self, data: u32) -> ProtoResult<()> {
-        self.buffer.reserve(4); // four bytes coming...
-
-        let b1: u8 = (data >> 24 & 0xFF) as u8;
-        let b2: u8 = (data >> 16 & 0xFF) as u8;
-        let b3: u8 = (data >> 8 & 0xFF) as u8;
-        let b4: u8 = (data & 0xFF) as u8;
-
-        self.emit(b1)?;
-        self.emit(b2)?;
-        self.emit(b3)?;
-        self.emit(b4)?;
-
+        let mut bytes = [0; 4];
+        {
+            NetworkEndian::write_u32(&mut bytes, data);
+        }
+        self.write_slice(&bytes);
         Ok(())
     }
 
+    fn write_slice(&mut self, data: &[u8]) {
+        // TODO: check whether this is better that letting `Vec.extend_from_slice` do the reservation.
```

Comment:
I don't see any difference  in the benchmarks, when I remove the explicit reservation, but the benchmarks are pretty limited, so it's hard to tell if this helps or not.